### PR TITLE
fix: body-transformer should have a smaller priority than proxy-cache

### LINKED ISSUE #9226: bug: body-transformer should have a smaller priority than proxy-cache
### Current Behavior

If we use the memory cache type in proxy-cache, then the body-transformer will not work.
Because currently, the proxy-rewrite runs the body filter after the body-transformer, so it caches the transformed body but not the original one from upstream. At the next request, the cache will hit and provide the transformed body to the body-transformer, it failed.

Refer to https://the-asf.slack.com/archives/CUC5MN17A/p1680504803770069 for use case.

### Expected Behavior

The proxy-cache should cache the original response from the upstream.

### Error Logs

2023/04/03 16:40:32 [error] 2705998#2705998: *279 [lua] body-transformer.lua:123: transform(): response body decode: /opt/apisix//deps/share/lua/5.1/xml2lua.lua:92: Err
or Parsing XML [char=1]
, body={"result": "hello"}, client: 127.0.0.1, server: _, request: "POST /capital HTTP/1.1", host: "127.0.0.1:9080"

### Steps to Reproduce

```bash
cat >/opt/go_http_server.go <<"EOF"
package main

import (
    "flag"
    "fmt"
    "log"
    "net/http"
)

var RESP string = `<SOAP-ENV:Envelope
    xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
    <SOAP-ENV:Header/>
    <SOAP-ENV:Body>
        <ns2:CapitalCityResponse
            xmlns:ns2="http://spring.io/guides/gs-producing-web-service">
            <ns2:CapitalCityResult>hello</ns2:CapitalCityResult>
        </ns2:CapitalCityResponse>
    </SOAP-ENV:Body>
</SOAP-ENV:Envelope>`

func main() {
    var listen = flag.String("listen", ":3000", "listen port")
    flag.Parse()

    // set up handler to listen to root path
    handler := http.NewServeMux()
    handler.HandleFunc("/", func(writer http.ResponseWriter, request *http.Request) {
        fmt.Fprintf(writer, RESP)
    })

    // serve on port 9090 of local host
    server := http.Server{
        Addr:    *listen,
        Handler: handler,
    }

    // serve the endpoint with tls encryption
    if err := server.ListenAndServe(); err != nil {
        log.Fatalf("error listening to port: %v", err)
    }
}
EOF

go run /opt/go_http_server.go

c put /routes/1 -d '
{
  "uri":"/capital",
  "name":"TempConvertRoute",
  "methods":[
    "GET",
    "POST",
    "PUT",
    "DELETE",
    "PATCH",
    "HEAD",
    "OPTIONS",
    "CONNECT",
    "TRACE",
    "PURGE"
  ],
  "plugins":{
    "body-transformer":{
      "request":{
        "input_format":"json",
        "template":"<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:web=\"http://www.oorsprong.org/websamples.countryinfo\"><soapenv:Header/><soapenv:Body><web:CapitalCity><web:sCountryISOCode>{{_escape_xml(country)}}</web:sCountryISOCode></web:CapitalCity></soapenv:Body></soapenv:Envelope>"
      },
      "response":{
        "input_format":"xml",
        "template":"{\"result\": {*_escape_json(Envelope.Body.CapitalCityResponse.CapitalCityResult)*}}"
      }
    },
    "file-logger":{
      "path":"logs/file.log"
    },
    "proxy-cache":{
      "cache_strategy":"memory",
      "cache_bypass":[
        "$arg_bypass"
      ],
      "cache_http_status":[
        200
      ],
      "cache_key":[
        "$uri",
        "-cache-id"
      ],
      "cache_method":[
        "POST"
      ],
      "hide_cache_headers":true,
      "no_cache":[
        "$arg_test"
      ],
      "cache_zone":"memory_cache"
    },
    "proxy-rewrite":{
      "headers":{
        "set":{
          "Accept-Encoding":"identity",
          "Content-Type":"text/xml"
        }
      },
      "uri":"/websamples.countryinfo/CountryInfoService.wso"
    },
    "response-rewrite":{
      "headers":{
        "set":{
          "Content-Type":"application/json"
        }
      }
    }
  },
  "upstream":{
    "nodes":{
      "localhost:3000":1
    }
  },
  "status":1
}
'

d post /capital -d '{"country":"foo"}' -v
```

### Environment

- APISIX version (run `apisix version`): master
- Operating system (run `uname -a`):
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`):

### PR DESCRIPTION
**Pull request by Sn0rt**
### Description

Fixes #9226 

this merge request base on #9266 

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

the test case is copy from the #9226 

```shell
$ prove -I. -I../test-nginx/inc -I../test-nginx/lib -r t/plugin/body-transformer.t
t/plugin/body-transformer.t .. ok
All tests successful.
Files=1, Tests=20, 14 wallclock secs ( 0.05 usr  0.00 sys +  1.97 cusr  0.58 csys =  2.60 CPU)
Result: PASS
```

### CODE REVIEW

File: t/plugin/body-transformer.t [hunk 1]
```diff
@@ -23,9 +23,28 @@ no_root_location();
 add_block_preprocessor(sub {
     my ($block) = @_;
 
-    if (!$block->request) {
-        $block->set_value("request", "GET /t");
+    my $http_config = $block->http_config // <<_EOC_;
+
+    # for proxy cache
```
Comment (kingluo) [inline t/plugin/body-transformer.t:28]: We use a memory-only cache handler, so do we need disk zone settings?

File: t/plugin/body-transformer.t [hunk 2]
```diff
@@ -652,3 +671,112 @@ foobar:
             assert(res.status == 200)
         }
     }
+
+
+=== TEST 10: cooperation of cache-proxy plugin
```
Comment (kingluo) [inline t/plugin/body-transformer.t:676]: typo fix: `cache-proxy` -> `proxy-cache`

File: t/plugin/body-transformer.t [hunk 3]
```diff
@@ -652,3 +662,113 @@ foobar:
             assert(res.status == 200)
         }
     }
+
+
+
+=== TEST 10: cooperation of proxy-cache plugin
+--- config
+location /websamples.countryinfo/CountryInfoService.wso {
```
Comment (monkeyDluffy6017) [inline t/plugin/body-transformer.t:670]: Why is this URI so long and does it have any special meaning?

**Comment by Sn0rt** [comment]
> Hi @Sn0rt, only a reviewer can mark a coversation as resolved

copy.

**Review by monkeyDluffy6017 — APPROVED** [review]

File: t/plugin/body-transformer.t [hunk 4]
```diff
@@ -23,9 +23,19 @@ no_root_location();
...
-    if (!$block->request) {
-        $block->set_value("request", "GET /t");
+    my $http_config = $block->http_config // <<_EOC_;
+
+    lua_shared_dict memory_cache 50m;
+
+    server {
+        listen 1986;
+        server_tokens off;
+
```
Comment (leslie-tsang) [inline t/plugin/body-transformer.t:33]: It seems we don't use this server ?

File: t/plugin/body-transformer.t [hunk 5]
```diff
@@ -652,3 +652,115 @@ foobar:
...
+                {"result": {*_escape_json(Envelope.Body.CapitalCityResponse.CapitalCityResult)*}}
+                ]]
+
+            local code, body = t.test('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                string.format([[{
+                    "uri": "/capital",
+                    "plugins": {
+                        "proxy-rewrite": {
+                            "set": {
```
Comment (soulbird) [inline t/plugin/body-transformer.t:704]: We can set these parameters when making a request, using the proxy-rewrite plugin seems to have nothing to do with the test target

**Review by soulbird — APPROVED** [review]

**Review by kingluo — APPROVED** [review]

**Review by monkeyDluffy6017 — APPROVED** [review]

**Pull request merged**